src/ore/src/stats.rs

```diff
@@ -75,3 +75,173 @@ pub const HISTOGRAM_BYTE_BUCKETS: [f64; 7] = [
     67108864.0,
     1073741824.0,
 ];
+
+/// Keeps track of the minimum and maximum value over a fixed-size sliding window of samples.
+///
+/// Inspired by the [`moving_min_max`] crate, see that crate's documentation for a description of
+/// the high-level algorithm used here.
+///
+/// There are two major differences to [`moving_min_max`]:
+///  * `SlidingMinMax` tracks both the minimum and the maximum value at the same time.
+///  * `SlidingMinMax` assumes a fixed-size window. Pushing new samples automatically pops old ones
+///    and there is no support for manually popping samples.
+///
+/// The memory required for a `SlidingMinMax` value is `size_of::<T> * 3 * window_size`, plus a
+/// small constant overhead.
+///
+/// [`moving_min_max`]: https://crates.io/crates/moving_min_max
+#[derive(Debug)]
+pub struct SlidingMinMax<T> {
+    /// The push stack and the pop stack, merged into one allocation to optimize memory usage.
+    ///
+    /// The push stack is the first `push_stack_len` items, the pop stack is the rest.
+    /// The push stack grows to the left, the pop stack grows to the right.
```

Comment:
Reading this more carefully, shouldn't this be the push stack grows to the right, pop stack to the left?
In the diagram below the arrows indicate that's what's happening

Reply:
Ah, good catch!

Reply:
🤦 Thanks!